### ISSUE TITLE
GM segmentation performance on shifted images

### ISSUE THREAD

**Issue by Nilser3** [ivadomed/model-gm-contrast-region-agnostic#7]
# Description 
In some cases the SC is off-center of the acquired image matrix, so it is necessary to test automatic segmentation models on images where the SC is off-center of the image (shifted image) and how robust they are to the shift.

# Proposition
With this [shift_image.py](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/shift_image.py) it is possible to move the center of an image in the axial plane in a radius ( `-r` : random range in **mm**), and with a random angulation. There is no shift in Z-axis.

The [shift_image.py](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/shift_image.py) can also be input segmentations `(-s1`, `-s2`) that will have the same shift as anat :

```sh

python shift_image.py -i anat.nii.gz -s1 anat_gmseg.nii.gz -s2 anat_contrast-agnostic-gm_seg.nii.gz  -r 35-50

```

![Image](https://github.com/user-attachments/assets/eab84bcf-4e3e-4b3a-87f3-7e33b74136d2)
In this case, the image was shifted by 48 mm (randomly from `-r `range) and an angle of 109 degrees (randomly).

The idea is to generate 5 areas or zones where the images will be displaced (each point is the center of a shifted subject).
- A0 : from 0 to 5 mm
- A1 : from 5 to 20 mm
- A2 : from 20 to 35 mm
- A3 : from 35 to 50 mm
- A4 : from 50 to 75 mm
- A5 : from 75 to 90 mm

<img src="https://github.com/user-attachments/assets/07e35f94-7577-43e8-b287-bb85f54c622a" alt="Image" width="500"/>

## Data
- 4 T2Star images from `sct-testing-large`, dimensions : 512 x 512 x 20 , resolutions: 0.3516 x 0.3516 x 3,3, that is, a dimension of 180 x 180 mm in axial-plane.
- The 4 anat images were automatically segmented (using `sct_deepseg_gm` SCT v6.5 and the [`contrast-region-agnostic-gm`](https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/2#issuecomment-2548138384) model 801) in their anat native space (no shifted), then this 3-pack was shifted 5 times in each area mentioned above (so we had 20 images in each Area).
- A re-segmentation was done on the shifted anat 
- The Dice Score was calculated with the shifted native masks, so we would expect to have Dice Scores close to 1 (since the same Method is applied in each case).

# Preliminary results 
1.`sct_deepseg_gm` STC v6.5
2. [`contrast-region-agnostic-gm`](https://github.com/ivadomed/model-gm-contrast-region-agnostic/issues/2#issuecomment-2548138384) model 801

<img src="https://github.com/user-attachments/assets/6768cc61-cfc1-47d3-a61c-3c81297f0d61" alt="Image" width="700"/>

It is observed that from a radial shift of 20 mm there is a drop in the Dice Score of the `sct_deepseg_gm` segmentation 
The `contrast-region-agnostic-gm` method is robust to shift, with a slight Dice decrease in cases of shift above 90 mm (really extreme cases). 

Tagging @naga-karthik and @jcohenadad 

# Related issues
[#130](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/130) #2

**Comment by jcohenadad** [comment]
> With this [shift_image.py](https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/shift_image.py) it is possible to move the center of an image in the axial plane in a radius ( -r : random range in mm), and with a random angulation. There is no shift in Z-axis.

I don't understand why there is a need to re-develop such a new function-- torchIO or other libraries should already have random translation and rotation, no? This seems to be pretty basic. @naga-karthik @tomDag25 what do you think?

**Comment by tomDag25** [comment]
I didn't try it with other libraries but with monai you can use the transforms : [RandRotated](https://docs.monai.io/en/stable/transforms.html#randrotated) to perform random rotation and [RandAffined](https://docs.monai.io/en/stable/transforms.html#randaffined) to perform random transformations like translation. I never used RandAffined but it looks like you can also include your rotation in this transform and therefore only use this one and not RandRotated. 

Note: the link I provided are the one for a dictionary usage, if you don't want to use a dictionary you can use these transforms instead: [RandRotate](https://docs.monai.io/en/stable/transforms.html#randrotate), [RandAffine](https://docs.monai.io/en/stable/transforms.html#randaffine).

**Comment by naga-karthik** [comment]
> torchIO or other libraries should already have random translation and rotation, no?

Yes, they do! AND, the segmentation models also use it (I can only speak for contrast-agnostic, I use RandAffine) with some translation too). I think the problem is that the translation transformation used during data augmentation was not enough, which is why, Nilser created his own function after "looking" at the original image and figure out how many pixels it should be shifted. 

In my case, the [RandomAffine](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/main/monai/transforms.py#L79) translate_range might be too small, so even this range becomes a hyperparameter to tweak!

**Comment by jcohenadad** [comment]
>  I think the problem is that the translation transformation used during data augmentation was not enough, which is why, Nilser created his own function after "looking" at the original image and figure out how many pixels it should be shifted.

Is that so @Nilser3 ? If that's the case, why not opening an issue on the existing library so they create the feature (instead of having community members create their own hack)-- this will be more useful for the community in the long run if a well-maintained library like torchIO/MONAI offers missing features--

**Comment by jcohenadad** [comment]
> In my case, the [RandomAffine](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/main/monai/transforms.py#L79) translate_range might be too small, so even this range becomes a hyperparameter to tweak!

@naga-karthik I'm not sure I understand: is it too small? and if so are you facing the same issue as Nilser? or it is not too small, but it just requires manual setting of the range, hence your comment about the hyperparameter to tweak.

**Comment by naga-karthik** [comment]
> I'm not sure I understand: is it too small?

Yes, i believe so, I set the the `translate_range` parameter to `[-0.1, 0.1]`. As per the definition in the monai docs, it randomly samples a value from the uniform distribution (-0.1, 0.1). I realize that these values are so small that we might not even see any translation. 

These values have to be set manually, so yes, this also becomes a hyperparameter.

**Comment by jcohenadad** [comment]
but if you set the range to, say, [-50, 50], then why do we need https://github.com/ivadomed/model-gm-contrast-region-agnostic/blob/nlm/first_model/codes/shift_image.py?

**Comment by naga-karthik** [comment]
Yes, exactly! that's my argument too! We would not need any test-time translations (like Nilser's `shift_image.py`) if the issue of "image not centered, hence model not working" could be solved by having a larger translation range (e.g. [-50, 50]) during training. 

I will try training a model with larger translation and see how it affects the results.
